domains/primitives/runtime/src/lib.rs

```diff
@@ -104,8 +105,16 @@ sp_api::decl_runtime_apis! {
 
         /// Returns the storage root after applying the extrinsic.
         fn apply_extrinsic_with_post_state_root(extrinsic: <Block as BlockT>::Extrinsic) -> Vec<u8>;
+    }
 
+    /// API used to construct extrincsic.
+    pub trait DomainExtrinsicApi<PNumber: Encode + Decode, PHash: Encode + Decode> {
```

Comment:
Make sense! We can move the `construct_set_code_extrinsic` back to `DomainCoreApi`. 

But for `construct_submit_core_bundle_extrinsics`, although it is actually only needed by the system domain, we need to implement it for every domain unfortunately to make the compiler pass.

It is because the `Block` generic type parameter in both `DomainBundles` and `SystemDomainApi` within `DomainBundleProcessor`. The `construct_submit_core_bundle_extrinsics` api force the `Block` must be the same in `DomainBundles` and `SystemDomainApi`, but it is impossible for the core domain, thus the workaround/hack this PR bring is to let the core domain also implement this api (although it is never used). 

And although it is quite a hack, perhaps we should move `construct_submit_core_bundle_extrinsics` to `DomainCoreApi` too, as we need the core domain also to implement this api to workaround this above problem.

Reply:
Moving submit_core into `DomainCoreApi` still obeys its design goal, we probably don't have to do this refactoring at present if it had to break the code which is well-defined. I'm reluctant to see it comes with some cost on the readability, especially since saving about 200 lines of code is not significant considering a good amount of them are the imports and generics which actually do not incur much thinking cost. I won't look into the other comments until we have a consensus on this one.